src/hotspot/share/runtime/thread.cpp

```diff
@@ -482,25 +482,49 @@ void Thread::check_for_dangling_thread_pointer(Thread *thread) {
 }
 #endif
 
-// Is the target JavaThread protected by the calling Thread:
+// Is the target JavaThread protected by the calling Thread
+// or by some other mechanism:
 bool Thread::is_JavaThread_protected(const JavaThread* p) {
-  Thread* thread = Thread::current();
-  if (thread == p || thread->is_VM_thread()) {
-    // Target JavaThread is self or access by the VMThread is always protected:
+  // Do the simplest check first:
+  if (SafepointSynchronize::is_at_safepoint()) {
+    // The target is protected since JavaThreads cannot exit
+    // while we're at a safepoint.
+    return true;
+  }
+
+  // Now make the simple checks based on who the caller is:
+  Thread* current_thread = Thread::current();
+  if (current_thread == p || Threads_lock->owner() == current_thread) {
+    // Target JavaThread is self or calling thread owns the Threads_lock.
+    // Second check is the same as Threads_lock->owner_is_self(),
+    // but we already have the current thread so check directly.
     return true;
   }
 
   // Check the ThreadsLists associated with the calling thread (if any)
   // to see if one of them protects the target JavaThread:
-  for (SafeThreadsListPtr* stlp = thread->_threads_list_ptr; stlp != NULL; stlp = stlp->previous()) {
+  for (SafeThreadsListPtr* stlp = current_thread->_threads_list_ptr;
+       stlp != NULL; stlp = stlp->previous()) {
     if (stlp->list()->includes(p)) {
       // The target JavaThread is protected by this ThreadsList:
       return true;
     }
   }
 
-  // The target JavaThread is not on this ThreadsList or any of its
-  // nested ThreadsLists (if any) so it is not safe to query:
+  // Use this debug code with -XX:+UseNewCode to diagnose locations that
+  // are missing a ThreadsListHandle or other protection mechanism:
+  // guarantee(!UseNewCode, "current_thread=" INTPTR_FORMAT " is not protecting p="
+  //           INTPTR_FORMAT, p2i(current_thread), p2i(p));
+  guarantee(false, "current_thread=" INTPTR_FORMAT " is not protecting p="
+            INTPTR_FORMAT, p2i(current_thread), p2i(p));
+
```

Comment:
This will be deleted before pushing - right?

Reply:
Yup. It's just for the current Tier[1-8] test cycle.